README.md
# kaocha-cloverage

<!-- badges -->
[![CircleCI](https://circleci.com/gh/lambdaisland/kaocha-cloverage.svg?style=svg)](https://circleci.com/gh/lambdaisland/kaocha-cloverage) [![cljdoc badge](https://cljdoc.org/badge/lambdaisland/kaocha-cloverage)](https://cljdoc.org/d/lambdaisland/kaocha-cloverage) [![Clojars Project](https://img.shields.io/clojars/v/lambdaisland/kaocha-cloverage.svg)](https://clojars.org/lambdaisland/kaocha-cloverage)
<!-- /badges -->

Kaocha plugin to get code coverage reports through [Cloverage](https://github.com/cloverage/cloverage).

<!-- opencollective -->

&nbsp;

<img align="left" src="https://github.com/lambdaisland/open-source/raw/master/artwork/lighthouse_readme.png">

&nbsp;

## Support Lambda Island Open Source

kaocha-cloverage is part of a growing collection of quality Clojure libraries and
tools released on the Lambda Island label. If you are using this project
commercially then you are expected to pay it forward by
[becoming a backer on Open Collective](http://opencollective.com/lambda-island#section-contribute),
so that we may continue to enjoy a thriving Clojure ecosystem.

&nbsp;

&nbsp;

<!-- /opencollective -->

## Installation

deps.edn

```
lambdaisland/kaocha-cloverage {:mvn/version "1.1.89"}
```

project.clj

```
[lambdaisland/kaocha-cloverage "1.1.89"]
```

## Usage

Add the dependency

``` clojure
;; deps.edn
{:paths [...]
 :deps {...}
 :aliases
 {:test
  {:extra-deps {lambdaisland/kaocha {...}
                lambdaisland/kaocha-cloverage {...}}}}}
```

Enable the plugin

- Command line

```
bin/kaocha --plugin cloverage
```

- Configuration file

``` clojure
;; tests.edn
#kaocha/v1
{:plugins [:kaocha.plugin/cloverage]}
```

This plugin adds a large amount of command line options to Kaocha. Run `bin/kaocha --plugin cloverage --help` to see them.

Alternatively Cloverage can be configured through `tests.edn`. Source paths specified in your tests suites will be automatically instrumented. Other Cloverage options can be specified under a top level `:cloverage/opts` key. Run `bin/kaocha --plugin cloverage --print-config` to see the current default values.

```
;; tests.edn
#kaocha/v1
{:plugins [kaocha.plugin/cloverage]
 :cloverage/opts
 {:ns-exclude-regex [],
  :text? false,
  :lcov? false,
  :high-watermark 80,
  :fail-threshold 0,
  :output "target/coverage",
  :low-watermark 50,
  :ns-regex [],
  :summary? true,
  :coveralls? false,
  :emma-xml? false,
  :html? true,
  :nop? false,
  :codecov? false}}
```

<!-- contributing -->
## Contributing

Everyone has a right to submit patches to kaocha-cloverage, and thus become a contributor.

Contributors MUST

- adhere to the [LambdaIsland Clojure Style Guide](https://nextjournal.com/lambdaisland/clojure-style-guide)
- write patches that solve a problem. Start by stating the problem, then supply a minimal solution. `*`
- agree to license their contributions as EPL 1.0.
- not break the contract with downstream consumers. `**`
- not break the tests.

Contributors SHOULD

- update the CHANGELOG and README.
- add tests for new functionality.

If you submit a pull request that adheres to these rules, then it will almost
certainly be merged immediately. However some things may require more
consideration. If you add new dependencies, or significantly increase the API
surface, then we need to decide if these changes are in line with the project's
goals. In this case you can start by [writing a pitch](https://nextjournal.com/lambdaisland/pitch-template),
and collecting feedback on it.

`*` This goes for features too, a feature needs to solve a problem. State the problem it solves, then supply a minimal solution.

`**` As long as this project has not seen a public release (i.e. is not on Clojars)
we may still consider making breaking changes, if there is consensus that the
changes are justified.
<!-- /contributing -->

<!-- license -->
## License

Copyright &copy; 2018-2020 Arne Brasseur and contributors

Available under the terms of the Eclipse Public License 1.0, see LICENSE.txt
<!-- /license -->

CHANGELOG.md
# Unreleased

## Added

## Fixed

## Changed

# 1.1.89 (2022-11-09 / c2b2dbf)

## Added

## Fixed

## Changed

- Bumped Cloverage dependency to `1.2.4`.

# 1.0.75 (2020-11-20 / 4c85364)

## Fixed

- Fix cljdoc build

# 1.0.72 (2020-11-20 / 20ea1aa)

## Fixed

- handle correctly test-ns-regex, converting strings into regexes and
  adding cli support for the option as `--cov-test-ns-regex`

# 1.0.63 (2020-08-21 / 83f3706)

## Changed

- Bump Cloverage to 1.2.0

# 1.0.56 (2020-08-13 / 304162e)

## Changed

- Switch back to vanilla Cloverage, and bump it to 1.1.3. This version no longer
  contains an aot-compiled tools.reader.

# 1.0-45 (2020-03-29 / 8642129)

## Fixed

- Get rid of clojure.tools.reader warnings, by switching to a Cloverage version
  that does not include an AOT compiled clojure.tools.reader.

# 0.0-41 (2019-09-25 / 0f12673)

## Fixed

- Support code using tagged literals ([#6](https://github.com/lambdaisland/kaocha-cloverage/pull/6))

# 0.0-32 (2019-03-19 / 8da7e96)

## Added

- Allow passing excluded calls to Cloverage, see also [Cloverage#242](https://github.com/cloverage/cloverage/pull/242)

## Changed

- Upgrade Cloverage to version 1.1.1

# 0.0-22 (2018-12-10 / ee13a86)

## Fixed

- Don't rely on Cloverage to find namespaces, this fixes compatibility with Cucumber.

# 0.0-18 (2018-12-05 / 9c8c629)

## Fixed

- Allow configuring ns-regex and ns-exclude-regex from tests.edn ([#2](https://github.com/lambdaisland/kaocha-cloverage/pull/2))

# 0.0-6 (2018-11-15 / 8308b88)

## Fixed

- Correctly specify Cloverage as a dependency

# 0.0-3 (2018-11-15 / 80c323a)

## Changed

- Initial release

test/unit/kaocha/plugin/cloverage_test.clj
(ns kaocha.plugin.cloverage-test
  (:refer-clojure :exclude [symbol])
  (:require [clojure.test :refer :all]
            [clojure.tools.cli :as cli]
            [kaocha.core-ext :refer :all]
            [kaocha.plugin :as plugin]
            [kaocha.plugin.cloverage :as cov]
            [slingshot.slingshot :refer [try+]]))

(require 'matcher-combinators.test)

(defmethod assert-expr 'thrown+? [msg form]
  ;; (is (thrown? m expr))
  (let [m (second form)
        body (nthnext form 2)]
    `(try+
      ~@body
      (do-report {:type :fail, :message ~msg,
                  :expected '~form, :actual nil})
      (catch ~m e#
        (do-report {:type :pass, :message ~msg,
                    :expected '~form, :actual e#})
        e#))))

(defn- update-config' [config cli-opts]
  (let [opts (-> cli-opts
                 (cli/parse-opts cov/cli-opts)
                 :options)]
    (-> {:kaocha/cli-options opts
         :cloverage/opts config}
        (cov/update-config)
        :cloverage/opts)))

(deftest update-config-test
  (testing "when no options given"
    (testing "it populates with defaults"
      (is (= cov/default-opts (:cloverage/opts (cov/update-config {}))))))

  (let [re->str (fn [re]
                  (is (regex? re))
                  (str re))]

    (testing "--cov-output"
      (is (match? {:output "foo/bar"}
                  (update-config' {}
                                  ["--cov-output" "foo/bar"])))

      (is (match? {:output "foo/bar"}
                  (update-config' {:output "bar/baz"}
                                  ["--cov-output" "foo/bar"]))))
    (testing "--cov-text"
      (is (match? {:text? false} (update-config' {} [])))
      (is (match? {:text? true} (update-config' {:text? true} [])))
      (is (match? {:text? false} (update-config' {:text? true}
                                                 ["--no-cov-text"])))
      (is (match? {:text? true} (update-config' {} ["--cov-text"]))))

    (testing "--cov-html"
      (is (match? {:html? true} (update-config' {} [])))
      (is (match? {:html? false} (update-config' {:html? false} [])))
      (is (match? {:html? true} (update-config' {:html? false}
                                                ["--cov-html"])))
      (is (match? {:html? false} (update-config' {} ["--no-cov-html"]))))

    (testing "--emma-xml"
      (is (match? {:emma-xml? false} (update-config' {} [])))
      (is (match? {:emma-xml? true} (update-config' {:emma-xml? true} [])))
      (is (match? {:emma-xml? false} (update-config' {:emma-xml? true}
                                                     ["--no-emma-xml"])))
      (is (match? {:emma-xml? true} (update-config' {} ["--emma-xml"]))))

    (testing "--lcov"
      (is (match? {:lcov? false} (update-config' {} [])))
      (is (match? {:lcov? true} (update-config' {:lcov? true} [])))
      (is (match? {:lcov? false} (update-config' {:lcov? true}
                                                 ["--no-lcov"])))
      (is (match? {:lcov? true} (update-config' {} ["--lcov"]))))

    (testing "--codecov"
      (is (match? {:codecov? false} (update-config' {} [])))
      (is (match? {:codecov? true} (update-config' {:codecov? true} [])))
      (is (match? {:codecov? false} (update-config' {:codecov? true}
                                                    ["--no-codecov"])))
      (is (match? {:codecov? true} (update-config' {} ["--codecov"]))))

    (testing "--coveralls"
      (is (match? {:coveralls? false} (update-config' {} [])))
      (is (match? {:coveralls? true} (update-config' {:coveralls? true} [])))
      (is (match? {:coveralls? false} (update-config' {:coveralls? true}
                                                      ["--no-coveralls"])))
      (is (match? {:coveralls? true} (update-config' {} ["--coveralls"]))))

    (testing "--cov-summary"
      (is (match? {:summary? true} (update-config' {} [])))
      (is (match? {:summary? false} (update-config' {:summary? false} [])))
      (is (match? {:summary? true} (update-config' {:summary? false}
                                                   ["--cov-summary"])))
      (is (match? {:summary? false} (update-config' {} ["--no-cov-summary"]))))

    (testing "--cov-fail-threshold PERCENT"
      (is (match? {:fail-threshold 0} (update-config' {} [])))
      (is (match? {:fail-threshold 20} (update-config' {:fail-threshold 20} [])))
      (is (match? {:fail-threshold 42} (update-config' {:fail-threshold 20} ["--cov-fail-threshold" "42"])))
      (is (match? {:fail-threshold 42} (update-config' {} ["--cov-fail-threshold" "42"]))))

    (testing "--cov-low-watermark PERCENT"
      (is (match? {:low-watermark 50} (update-config' {} [])))
      (is (match? {:low-watermark 20} (update-config' {:low-watermark 20} [])))
      (is (match? {:low-watermark 42} (update-config' {:low-watermark 20} ["--cov-low-watermark" "42"])))
      (is (match? {:low-watermark 42} (update-config' {} ["--cov-low-watermark" "42"]))))

    (testing "--cov-high-watermark PERCENT"
      (is (match? {:high-watermark 80} (update-config' {} [])))
      (is (match? {:high-watermark 20} (update-config' {:high-watermark 20} [])))
      (is (match? {:high-watermark 42} (update-config' {:high-watermark 20} ["--cov-high-watermark" "42"])))
      (is (match? {:high-watermark 42} (update-config' {} ["--cov-high-watermark" "42"]))))

    (testing "--nop"
      (is (match? {:nop? false} (update-config' {} [])))
      (is (match? {:nop? true} (update-config' {:nop? true} [])))
      (is (match? {:nop? false} (update-config' {:nop? true}
                                                ["--no-cov-nop"])))
      (is (match? {:nop? true} (update-config' {} ["--cov-nop"]))))

    (testing "--cov-ns-regex"
      (is (= [] (:ns-regex (update-config' {} []))))
      (is (= ["foo.*"] (map re->str (:ns-regex (update-config' {:ns-regex ["foo.*"]} [])))))
      (is (= ["bar"] (map re->str (:ns-regex (update-config' {:ns-regex ["foo"]} ["--cov-ns-regex" "bar"]))))))

    (testing "--cov-src-ns-path"
      (is (= [] (:src-ns-path (update-config' {} []))))
      (is (= ["src"] (:src-ns-path (update-config' {:src-ns-path ["src"]} []))))
      (is (= ["src"] (:src-ns-path (update-config' {} ["--cov-src-ns-path" "src"])))))

    (testing "--cov-ns-exclude-regex"
      (is (= [] (:ns-exclude-regex (update-config' {} []))))
      (is (= ["foo.*"] (map re->str (:ns-exclude-regex (update-config' {:ns-exclude-regex ["foo.*"]} [])))))
      (is (= ["foo" "bar"] (map re->str (:ns-exclude-regex (update-config' {:ns-exclude-regex ["foo"]} ["--cov-ns-exclude-regex" "bar"]))))))

    (testing "--cov-exclude-call"
      (is (= [] (:exclude-call (update-config' {} []))))
      (is (= ['my.ns/fn] (:exclude-call (update-config' {:exclude-call ['my.ns/fn]} []))))
      (is (= ['my.other.ns/fn] (:exclude-call (update-config' {} ["--cov-exclude-call" "my.other.ns/fn"]))))
      (is (= ['my.ns/fn 'my.other.ns/fn] (:exclude-call (update-config' {:exclude-call ['my.ns/fn]} ["--cov-exclude-call" "my.other.ns/fn"])))))

    (testing "--test-ns-regex"
      (is (= [] (:test-ns-regex (update-config' {} []))))
      (is (= ["foo.*"] (map re->str (:test-ns-regex (update-config' {:test-ns-regex ["foo.*"]} [])))))
      (is (= ["foo" "bar"] (map re->str (:test-ns-regex (update-config' {:test-ns-regex ["foo"]} ["--cov-test-ns-regex" "bar"]))))))))

(deftest run-cloverage-test
  (let [arglists (:arglists (meta #'cloverage.coverage/run-main))]
    (try
      (let [run-main (fn [a1] a1)]
        (alter-meta! #'cloverage.coverage/run-main assoc :arglists '([:a1]))
        (with-redefs [cloverage.coverage/run-main run-main]
          (is (= [:opts] (cov/run-cloverage :opts)))))

      (let [run-main (fn [a1 a2] [a1 a2])]
        (alter-meta! #'cloverage.coverage/run-main assoc :arglists '([:a1 :a2]))
        (with-redefs [cloverage.coverage/run-main run-main]
          (is (= [[:opts] {}] (cov/run-cloverage :opts)))))

      (testing "no matching clause"
        (alter-meta! #'cloverage.coverage/run-main assoc :arglists '([:a1 :a2 :a3]))
        (is (thrown? java.lang.IllegalArgumentException (cov/run-cloverage :opts))))

      (finally
        (alter-meta! #'cloverage.coverage/run-main assoc :arglists arglists)))))

(deftest cloverage-plugin-test
  (let [chain (plugin/load-all [:kaocha.plugin/cloverage])]

    (testing "cli-options"
      (is (= [["-h" "--help" "print help"]
              [nil "--cov-output PATH" "Cloverage output directory."]]
             (take 2 (plugin/run-hook* chain :kaocha.hooks/cli-options [["-h" "--help" "print help"]])))))

    (testing "config"
      (is (contains? (plugin/run-hook* chain :kaocha.hooks/config {})
                     :cloverage/opts)))

    (with-redefs [kaocha.api/run (fn [config]
                                   {:kaocha.result/tests [{:kaocha.result/count 3
                                                           :kaocha.result/pass 1
                                                           :kaocha.result/fail 2}]})
                  ;; unfortunately we can't invoke Cloverage "for real", or it
                  ;; will really mess up our coverage results
                  cloverage.coverage/run-main (fn [[opts] & _]
                                                ((cloverage.coverage/runner-fn {:runner :kaocha}) opts))]

      (is (thrown+? :kaocha/early-exit
                    (plugin/run-hook* chain :kaocha.hooks/main {:cloverage/opts {:src-ns-path ["src"]}})))

      (is (thrown+? :kaocha/early-exit
                    (plugin/run-hook* chain
                                      :kaocha.hooks/main
                                      (kaocha.config/normalize {:tests [{:source-paths ["src"]}]})))))))


src/kaocha/plugin/cloverage.clj
(ns kaocha.plugin.cloverage
  (:require [cloverage.coverage :as c]
            [clojure.tools.reader :as reader]
            [kaocha.api :as api]
            [kaocha.plugin :as plugin :refer [defplugin]]
            [kaocha.result :as result]
            [kaocha.classpath :as cp]
            [slingshot.slingshot :refer [throw+]]))

(defn accumulate [m k v]
  (update m k (fnil conj []) v))

(def default-opts
  {:output "target/coverage"
   :text? false
   :html? true
   :emma-xml? false
   :lcov? false
   :codecov? false
   :coveralls? false
   :summary? true
   :fail-threshold 0
   :low-watermark 50
   :high-watermark 80
   :nop? false
   :src-ns-path []
   :ns-regex []
   :ns-exclude-regex []
   :exclude-call []
   :test-ns-regex []})

(def cli-opts*
  [["--cov-output PATH" "Cloverage output directory."]
   ["--[no-]cov-text" "Produce a text report."]
   ["--[no-]cov-html" "Produce a HTML report."]
   ["--[no-]emma-xml"
    "Produce an EMMA XML report. [emma.sourceforge.net]"]
   ["--[no-]lcov"
    "Produce a lcov/gcov report."]
   ["--[no-]codecov"
    "Generate a JSON report for Codecov.io"]
   ["--[no-]coveralls"
    "Send a JSON report to Coveralls if on a CI server"]
   ["--[no-]cov-summary"
    "Prints a summary"]
   ["--cov-fail-threshold PERCENT"
    "Sets the percentage threshold at which cloverage will abort the build. Default: 0%"
    :parse-fn #(Integer/parseInt %)]
   ["--cov-low-watermark PERCENT"
    "Sets the low watermark percentage (valid values 0..100). Default: 50%"
    :parse-fn #(Integer/parseInt %)]
   ["--cov-high-watermark PERCENT"
    "Sets the high watermark percentage (valid values 0..100). Default: 80%"
    :parse-fn #(Integer/parseInt %)]
   ["--[no-]cov-nop" "Instrument with noops."]
   ["--cov-test-ns-regex REGEX"
    "Regex for test namespaces (can be repeated)."
    :assoc-fn accumulate]
   ["--cov-ns-regex REGEX"
    "Regex for instrumented namespaces (can be repeated)."
    :assoc-fn accumulate]
   ["--cov-ns-exclude-regex REGEX"
    "Regex for namespaces not to be instrumented (can be repeated)."
    :assoc-fn accumulate]
   ["--cov-src-ns-path PATH"
    "Path (string) to directory containing test namespaces (can be repeated). Defaults to test suite source paths."
    :assoc-fn accumulate]
   ["--cov-exclude-call NAME"
    "Name of fn/macro whose call sites are not to be instrumented (can be repeated)."
    :assoc-fn accumulate]])

(def cli-opts (map #(into [nil] %) cli-opts*))

(defn update-config [config]
  (let [opts (:kaocha/cli-options config)]
    (update config
            :cloverage/opts
            #(cond-> (merge default-opts %)
               (contains? opts :cov-output)
               (assoc :output (:cov-output opts))

               (contains? opts :cov-text)
               (assoc :text? (:cov-text opts))

               (contains? opts :cov-html)
               (assoc :html? (:cov-html opts))

               (contains? opts :emma-xml)
               (assoc :emma-xml? (:emma-xml opts))

               (contains? opts :lcov)
               (assoc :lcov? (:lcov opts))

               (contains? opts :codecov)
               (assoc :codecov? (:codecov opts))

               (contains? opts :coveralls)
               (assoc :coveralls? (:coveralls opts))

               (contains? opts :cov-summary)
               (assoc :summary? (:cov-summary opts))

               (contains? opts :cov-fail-threshold)
               (assoc :fail-threshold (:cov-fail-threshold opts))

               (contains? opts :cov-low-watermark)
               (assoc :low-watermark (:cov-low-watermark opts))

               (contains? opts :cov-high-watermark)
               (assoc :high-watermark (:cov-high-watermark opts))

               (contains? opts :cov-nop)
               (assoc :nop? (:cov-nop opts))

               (contains? opts :cov-src-ns-path)
               (assoc :src-ns-path (:cov-src-ns-path opts))

               (contains? opts :cov-exclude-call)
               (update :exclude-call into (map symbol (:cov-exclude-call opts)))

               (contains? opts :cov-ns-regex)
               (assoc :ns-regex (:cov-ns-regex opts))

               (contains? opts :cov-ns-exclude-regex)
               (update :ns-exclude-regex into (:cov-ns-exclude-regex opts))

               (contains? opts :cov-test-ns-regex)
               (update :test-ns-regex into (:cov-test-ns-regex opts))

               :always
               (update :test-ns-regex (partial mapv re-pattern))

               :always
               (update :ns-regex (partial mapv re-pattern))

               :always
               (update :ns-exclude-regex (partial mapv re-pattern))))))

(defn run-cloverage [opts]
  ;; Compatibility with future versions
  (let [arity      (count (first (:arglists (meta #'c/run-main))))
        decls      (-> []
                       (.getClass)
                       (.getClassLoader)
                       (.getResources "data_readers.clj")
                       enumeration-seq)
        read-decls (comp read-string slurp)
        readers    (reduce merge {} (map read-decls decls))]
    (binding [reader/*data-readers* readers]
      (case arity
        1 (c/run-main [opts])
        2 (c/run-main [opts] {})))))

(defplugin kaocha.plugin/cloverage
  (cli-options [opts]
    (into opts cli-opts))

  ;; The config hook only gets called inside api/run, which is too late for us,
  ;; so update-config also gets called explicitly in `main`. Calling
  ;; update-config here anyway so people can see the result of option merging
  ;; with --print-config. It's a bit of duplicated work though, we might have to
  ;; rethink this to make people implementing custom `main` hooks able to still
  ;; use `config` hooks.

  (config [config]
    (update-config config))

  (main [config]
    (binding [c/*exit-after-test* false]
      (let [config         (update-config config)
            suites         (remove :kaocha.testable/skip (:kaocha/tests config))
            source-paths   (mapcat :kaocha/source-paths suites)
            test-paths     (mapcat :kaocha/test-paths suites)
            tests          (map :kaocha.testable/id suites)
            cloverage-opts (:cloverage/opts config)
            opts           (assoc cloverage-opts
                                  :runner :kaocha
                                  :src-ns-path (if-let [src-ns-path (seq (:src-ns-path cloverage-opts))]
                                                 src-ns-path
                                                 source-paths)
                                  :extra-test-ns tests
                                  ;;:debug? true
                                  ::config config)]
        (run! cp/add-classpath test-paths)
        (throw+ {:kaocha/early-exit (run-cloverage opts)})))))

(defmethod c/runner-fn :kaocha [opts]
  (fn [_test-nses]
    (let [result (->> (::config opts)
                      api/run
                      (plugin/run-hook :kaocha.hooks/post-summary)
                      :kaocha.result/tests
                      result/totals)]
      {:errors (+ (::result/error result)
                  (::result/fail result))})))
